include/RAJA/pattern/nested.hpp

```diff
@@ -0,0 +1,228 @@
+#ifndef RAJA_pattern_nested_HPP
+#define RAJA_pattern_nested_HPP
+
+#include "RAJA/RAJA.hpp"
+#include "RAJA/config.hpp"
+#include "RAJA/internal/tuple.hpp"
+#include "RAJA/util/defines.hpp"
+#include "RAJA/util/types.hpp"
+
+#include "RAJA/external/metal.hpp"
+
+#include <iostream>
+#include <type_traits>
+
+namespace RAJA
+{
+namespace nested
+{
+
+namespace internal
+{
+
+template <typename T>
+using remove_all_t =
+    typename std::remove_cv<typename std::remove_reference<T>::type>::type;
+
+// Universal base of all For wrappers for type traits
+struct ForList {
+};
+struct ForBase {
+};
+struct CollapseBase {
+};
+template <metal::int_ ArgumentId, typename IndexType = RAJA::Index_type>
+struct ForTraitBase : public ForBase {
+  using index_type = IndexType;
+  using index_pair = metal::pair<metal::number<ArgumentId>, IndexType>;
+  using index = metal::number<ArgumentId>;
+};
+
+using is_for_policy = metal::
+    bind<metal::trait<std::is_base_of>, metal::always<ForBase>, metal::_1>;
+
+using has_for_list = metal::
+    bind<metal::trait<std::is_base_of>, metal::always<ForList>, metal::_1>;
+template<typename T>
+using get_for_list = typename T::as_for_list;
+
+template <typename Seq>
+using get_for_policies = metal::flatten<
+    metal::transform<
+        metal::lambda<get_for_list>,
+        metal::remove_if<
+            Seq,
+            metal::bind<metal::lambda<metal::not_>, internal::has_for_list>>>>;
+
+template <typename ForPolicy>
+using get_for_index_pair = typename ForPolicy::index_pair;
+
+template <typename PairL, typename PairR>
+using order_by_index =
+    metal::number<(metal::first<PairL>::value < metal::first<PairR>::value)>;
+
+template <typename Seq>
+using get_for_policies_as_pairs =
+    metal::transform<metal::lambda<get_for_index_pair>, get_for_policies<Seq>>;
+
+template <typename Seq>
+using get_for_index_types = metal::transform<
+    metal::lambda<metal::second>,
+    metal::sort<get_for_policies_as_pairs<Seq>, metal::lambda<order_by_index>>>;
+
+template <typename Seq>
+using index_tuple_from_policies =
+    metal::apply<metal::lambda<RAJA::util::tuple>, get_for_index_types<Seq>>;
+}
+
+template <metal::int_ ArgumentId,
+          typename Pol,
+          typename IndexType = RAJA::Index_type>
+struct For : public internal::ForList,
+             public internal::ForTraitBase<ArgumentId, IndexType> {
+  using as_for_list = metal::list<For<ArgumentId, Pol, IndexType>>;
+  // TODO: add static_assert for valid policy in Pol
```

Comment:
When you say valid policy, will we know the previously applied policies as well? For example, omp_parallel_for_exec within a cuda_exec makes for a bad time. We could theoretically check for that :)